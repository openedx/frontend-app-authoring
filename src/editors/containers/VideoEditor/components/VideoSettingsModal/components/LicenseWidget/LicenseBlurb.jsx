import React from 'react';
import PropTypes from 'prop-types';
import {
  FormattedMessage,
} from '@edx/frontend-platform/i18n';
import { Icon } from '@openedx/paragon';
import {
  Attribution,
  Copyright,
  Cc,
  Nd,
  Nc,
  Sa,
} from '@openedx/paragon/icons';

import messages from './messages';
import { LicenseTypes } from '../../../../../../data/constants/licenses';

const LicenseBlurb = ({
  license,
  details,
}) => (
  <div className="d-flex flex-row align-items-center mt-2">
    {/* not sure how to handle the edge cases when some of the icons are not displayed */}
    {license === LicenseTypes.allRightsReserved ? <Icon src={Copyright} className="mr-1" style={{ height: '18px', width: '18px' }} /> : null}
    {license === LicenseTypes.creativeCommons ? <Icon src={Cc} className="mr-1" style={{ height: '18px', width: '18px' }} /> : null}
    {details.attribution ? <Icon src={Attribution} className="mr-1 text-primary-300" style={{ height: '18px', width: '18px' }} /> : null}
    {details.noncommercial ? <Icon src={Nc} className="mr-1" style={{ height: '18px', width: '18px' }} /> : null}
    {details.noDerivatives ? <Icon src={Nd} className="mr-1" style={{ height: '18px', width: '18px' }} /> : null}
    {details.shareAlike ? <Icon src={Sa} className="mr-1" style={{ height: '18px', width: '18px' }} /> : null}
    {license === LicenseTypes.allRightsReserved
      ? <div className="small mx-1.5"><FormattedMessage {...messages.allRightsReservedIconsLabel} /></div>
      : null}
    {license === LicenseTypes.creativeCommons
      ? <div className="small mx-1.5"><FormattedMessage {...messages.creativeCommonsIconsLabel} /></div>
      : null}
  </div>
);

LicenseBlurb.propTypes = {
  license: PropTypes.string.isRequired,
  details: PropTypes.shape({
    attribution: PropTypes.bool.isRequired,
    noncommercial: PropTypes.bool.isRequired,
    noDerivatives: PropTypes.bool.isRequired,
    shareAlike: PropTypes.bool.isRequired,
  }).isRequired,
};

export default LicenseBlurb;
