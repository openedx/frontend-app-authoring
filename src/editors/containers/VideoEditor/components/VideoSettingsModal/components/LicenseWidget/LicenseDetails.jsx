import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  FormattedMessage,
  injectIntl,
} from '@edx/frontend-platform/i18n';
import {
  ActionRow,
  CheckboxControl,
  Form,
  Icon,
  Stack,
} from '@openedx/paragon';
import {
  Attribution,
  Nd,
  Sa,
  Nc,
} from '@openedx/paragon/icons';

import { actions } from '../../../../../../data/redux';
import { LicenseLevel, LicenseTypes } from '../../../../../../data/constants/licenses';
import messages from './messages';

const LicenseDetails = ({
  license,
  details,
  level,
  // redux
  updateField,
}) => (
  level !== LicenseLevel.course && details && license !== 'select' ? (
    <div className="x-small border-primary-100 border-bottom m-0 pr-1">
      <Form.Group className="pb-2">
        <div className="mb-3">
          <FormattedMessage {...messages.detailsSubsectionTitle} />
        </div>
        {license === LicenseTypes.allRightsReserved ? (
          <div className="mt-2">
            <FormattedMessage {...messages.allRightsReservedSectionMessage} />
          </div>
        ) : null}
        {license === LicenseTypes.creativeCommons
          ? (
            <Stack gap={4}>
              <div className="border-primary-100 border-bottom pb-4">
                <Form.Group>
                  <ActionRow>
                    <Icon className="text-primary-500" src={Attribution} />
                    <Form.Label className="my-0 text-primary-500">
                      <FormattedMessage {...messages.attributionCheckboxLabel} />
                    </Form.Label>
                    <ActionRow.Spacer />
                    <CheckboxControl
                      disabled
                      checked
                      aria-label="Checkbox"
                    />
                  </ActionRow>
                </Form.Group>
                <div>
                  <FormattedMessage {...messages.attributionSectionDescription} />
                </div>
              </div>
              <div className="border-primary-100 border-bottom pb-4">
                <Form.Group>
                  <ActionRow>
                    <Icon src={Nc} />
                    <Form.Label className="my-0 text-primary-500">
                      <FormattedMessage {...messages.noncommercialCheckboxLabel} />
                    </Form.Label>
                    <ActionRow.Spacer />
                    <CheckboxControl
                      checked={details.noncommercial}
                      disabled={level === LicenseLevel.course}
                      onChange={(e) => updateField({
                        licenseDetails: {
                          ...details,
                          noncommercial: e.target.checked,
                        },
                      })}
                      aria-label="Checkbox"
                    />
                  </ActionRow>
                </Form.Group>
                <div>
                  <FormattedMessage {...messages.noncommercialSectionDescription} />
                </div>
              </div>
              <div className="border-primary-100 border-bottom pb-4">
                <Form.Group>
                  <ActionRow>
                    <Icon src={Nd} />
                    <Form.Label className="my-0 text-primary-500">
                      <FormattedMessage {...messages.noDerivativesCheckboxLabel} />
                    </Form.Label>
                    <ActionRow.Spacer />
                    <CheckboxControl
                      checked={details.noDerivatives}
                      disabled={level === LicenseLevel.course}
                      onChange={(e) => updateField({
                        licenseDetails: {
                          ...details,
                          noDerivatives: e.target.checked,
                          shareAlike: e.target.checked ? false : details.shareAlike,
                        },
                      })}
                      aria-label="Checkbox"
                    />
                  </ActionRow>
                </Form.Group>
                <div>
                  <FormattedMessage {...messages.noDerivativesSectionDescription} />
                </div>
              </div>
              <div>
                <Form.Group>
                  <ActionRow>
                    <Icon src={Sa} />
                    <Form.Label className="my-0 text-primary-500">
                      <FormattedMessage {...messages.shareAlikeCheckboxLabel} />
                    </Form.Label>
                    <ActionRow.Spacer />
                    <CheckboxControl
                      checked={details.shareAlike}
                      disabled={level === LicenseLevel.course}
                      onChange={(e) => updateField({
                        licenseDetails: {
                          ...details,
                          shareAlike: e.target.checked,
                          noDerivatives: e.target.checked ? false : details.noDerivatives,
                        },
                      })}
                      aria-label="Checkbox"
                    />
                  </ActionRow>
                </Form.Group>
                <div>
                  <FormattedMessage {...messages.shareAlikeSectionDescription} />
                </div>
              </div>
            </Stack>
          ) : null}
      </Form.Group>
    </div>
  ) : null
);

LicenseDetails.propTypes = {
  license: PropTypes.string.isRequired,
  details: PropTypes.shape({
    attribution: PropTypes.bool.isRequired,
    noncommercial: PropTypes.bool.isRequired,
    noDerivatives: PropTypes.bool.isRequired,
    shareAlike: PropTypes.bool.isRequired,
  }).isRequired,
  level: PropTypes.string.isRequired,
  // redux
  updateField: PropTypes.func.isRequired,
};

export const mapStateToProps = () => ({});

export const mapDispatchToProps = (dispatch) => ({
  updateField: (stateUpdate) => dispatch(actions.video.updateField(stateUpdate)),
});

export const LicenseDetailsInternal = LicenseDetails; // For testing only
export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(LicenseDetails));
