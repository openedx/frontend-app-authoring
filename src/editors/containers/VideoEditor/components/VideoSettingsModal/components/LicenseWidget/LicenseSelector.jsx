import React from 'react';
import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  FormattedMessage,
  useIntl,
} from '@edx/frontend-platform/i18n';
import {
  ActionRow,
  Form,
  Icon,
  IconButtonWithTooltip,
} from '@openedx/paragon';
import { DeleteOutline } from '@openedx/paragon/icons';

import { actions, selectors } from '../../../../../../data/redux';
import * as hooks from './hooks';
import messages from './messages';
import { LicenseLevel, LicenseNames, LicenseTypes } from '../../../../../../data/constants/licenses';

const LicenseSelector = ({
  license,
  level,
  // redux
  courseLicenseType,
  updateField,
}) => {
  const intl = useIntl();
  const { levelDescription } = hooks.determineText({ level });
  const onLicenseChange = hooks.onSelectLicense({ dispatch: useDispatch() });
  const ref = React.useRef();
  return (
    <>
      <ActionRow>
        <Form.Control
          as="select"
          className="w-100 m-0 p-0"
          ref={ref}
          defaultValue={license}
          disabled={level === LicenseLevel.course}
          floatingLabel={intl.formatMessage(messages.licenseTypeLabel)}
          onChange={(e) => onLicenseChange(e.target.value)}
        >
          {Object.entries(LicenseNames).map(([key, text]) => {
            if (license === key) { return (<option value={LicenseTypes[key]} selected>{text}</option>); }
            if (key === LicenseTypes.select) { return (<option hidden>{text}</option>); }
            return (<option value={LicenseTypes[key]}>{text}</option>);
          })}
        </Form.Control>
        {level !== LicenseLevel.course ? (
          <>
            <ActionRow.Spacer />
            <IconButtonWithTooltip
              iconAs={Icon}
              src={DeleteOutline}
              onClick={() => {
                ref.current.value = courseLicenseType;
                updateField({ licenseType: '', licenseDetails: {} });
              }}
              tooltipPlacement="top"
              tooltipContent={<FormattedMessage {...messages.deleteLicenseSelection} />}
            />
          </>
        ) : null }
      </ActionRow>
      <div className="x-small mt-3">{levelDescription}</div>
      {license === LicenseTypes.select ? null : <div className="border-primary-100 mt-3 border-bottom" />}
    </>
  );
};

LicenseSelector.propTypes = {
  license: PropTypes.string.isRequired,
  level: PropTypes.string.isRequired,
  // redux
  courseLicenseType: PropTypes.string.isRequired,
  updateField: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  courseLicenseType: selectors.video.courseLicenseType(state),
});

export const mapDispatchToProps = (dispatch) => ({
  updateField: (stateUpdate) => dispatch(actions.video.updateField(stateUpdate)),
});

export const LicenseSelectorInternal = LicenseSelector; // For testing only
export default connect(mapStateToProps, mapDispatchToProps)(LicenseSelector);
