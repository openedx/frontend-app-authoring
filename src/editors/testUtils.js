/* istanbul ignore file */
import react from 'react';
import { StrictDict } from './utils';

/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};

/**
 * Mock a single component, or a nested component so that its children render nicely
 * in snapshots.
 * @param {string} name - parent component name
 * @param {obj} contents - object of child components with intended component
 *   render name.
 * @return {func} - mock component with nested children.
 *
 * usage:
 *   mockNestedComponent('Card', { Body: 'Card.Body', Form: { Control: { Feedback: 'Form.Control.Feedback' }}... });
 *   mockNestedComponent('IconButton', 'IconButton');
 */
export const mockNestedComponent = (name, contents) => {
  if (typeof contents !== 'object') {
    return contents;
  }
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(contents).forEach((nestedName) => {
    const value = contents[nestedName];
    fn[nestedName] = typeof value !== 'object'
      ? value
      : mockNestedComponent(`${name}.${nestedName}`, value);
  });
  return fn;
};

/**
 * Mock a module of components.  nested components will be rendered nicely in snapshots.
 * @param {obj} mapping - component module mock config.
 * @return {obj} - module of flat and nested components that will render nicely in snapshots.
 * usage:
 *   mockNestedComponents({
 *     Card: { Body: 'Card.Body' },
 *     IconButton: 'IconButton',
 *   })
 */
export const mockNestedComponents = (mapping) => Object.entries(mapping).reduce(
  (obj, [name, value]) => ({
    ...obj,
    [name]: mockNestedComponent(name, value),
  }),
  {},
);

/**
 * Mock utility for working with useState in a hooks module.
 * Expects/requires an object containing the state object in order to ensure
 * the mock behavior works appropriately.
 *
 * Expected format:
 *   hooks = { state: { <key>: (val) => React.createRef(val), ... } }
 *
 * Returns a utility for mocking useState and providing access to specific state values
 * and setState methods, as well as allowing per-test configuration of useState value returns.
 *
 * Example usage:
 *   // hooks.js
 *   import * as module from './hooks';
 *   const state = {
 *     isOpen: (val) => React.useState(val),
 *     hasDoors: (val) => React.useState(val),
 *     selected: (val) => React.useState(val),
 *   };
 *   ...
 *   export const exampleHook = () => {
 *     const [isOpen, setIsOpen] = module.state.isOpen(false);
 *     if (!isOpen) { return null; }
 *     return { isOpen, setIsOpen };
 *   }
 *   ...
 *
 *   // hooks.test.js
 *   import * as hooks from './hooks';
 *   const state = new MockUseState(hooks)
 *   ...
 *   describe('state hooks', () => {
 *     state.testGetter(state.keys.isOpen);
 *     state.testGetter(state.keys.hasDoors);
 *     state.testGetter(state.keys.selected);
 *   });
 *   describe('exampleHook', () => {
 *     beforeEach(() => { state.mock(); });
 *     it('returns null if isOpen is default value', () => {
 *       expect(hooks.exampleHook()).toEqual(null);
 *     });
 *     it('returns isOpen and setIsOpen if isOpen is not null', () => {
 *       state.mockVal(state.keys.isOpen, true);
 *       expect(hooks.exampleHook()).toEqual({
 *         isOpen: true,
 *         setIsOpen: state.setState[state.keys.isOpen],
 *       });
 *     });
 *     afterEach(() => { state.restore(); });
 *   });
 *
 * @param {obj} hooks - hooks module containing a 'state' object
 */
export class MockUseState {
  constructor(hooks) {
    this.hooks = hooks;
    this.oldState = null;
    this.setState = {};
    this.stateVals = {};

    this.mock = this.mock.bind(this);
    this.restore = this.restore.bind(this);
    this.mockVal = this.mockVal.bind(this);
    this.testGetter = this.testGetter.bind(this);
  }

  /**
   * @return {object} - StrictDict of state object keys
   */
  get keys() {
    return StrictDict(Object.keys(this.hooks.state).reduce(
      (obj, key) => ({ ...obj, [key]: key }),
      {},
    ));
  }

  /**
   * Replace the hook module's state object with a mocked version, initialized to default values.
   */
  mock() {
    this.oldState = this.hooks.state;
    Object.keys(this.keys).forEach(key => {
      this.hooks.state[key] = jest.fn(val => {
        this.stateVals[key] = val;
        return [val, this.setState[key]];
      });
    });
    this.setState = Object.keys(this.keys).reduce(
      (obj, key) => ({
        ...obj,
        [key]: jest.fn(val => {
          this.hooks.state[key] = val;
        }),
      }),
      {},
    );
  }

  /**
   * Restore the hook module's state object to the actual code.
   */
  restore() {
    this.hooks.state = this.oldState;
  }

  /**
   * Mock the state getter associated with a single key to return a specific value one time.
   * @param {string} key - state key (from this.keys)
   * @param {any} val - new value to be returned by the useState call.
   */
  mockVal(key, val) {
    this.hooks.state[key].mockReturnValueOnce([val, this.setState[key]]);
  }

  testGetter(key) {
    test(`${key} state getter should return useState passthrough`, () => {
      const testValue = 'some value';
      const useState = (val) => ({ useState: val });
      jest.spyOn(react, 'useState').mockImplementationOnce(useState);
      expect(this.hooks.state[key](testValue)).toEqual(useState(testValue));
    });
  }
}
