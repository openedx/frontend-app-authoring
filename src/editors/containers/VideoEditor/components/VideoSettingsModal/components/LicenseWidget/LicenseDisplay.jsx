import React from 'react';
import PropTypes from 'prop-types';

import {
  FormattedMessage,
} from '@edx/frontend-platform/i18n';
import {
  Stack,
  Hyperlink,
} from '@openedx/paragon';

import { LicenseTypes } from '../../../../../../data/constants/licenses';

import LicenseBlurb from './LicenseBlurb';
import messages from './messages';

const LicenseDisplay = ({
  license,
  details,
  licenseDescription,
}) => {
  if (license !== LicenseTypes.select) {
    return (
      <Stack gap={3}>
        <div className="x-small"><FormattedMessage {...messages.displaySubsectionTitle} /></div>
        <div className="small border border-gray-300 rounded p-4">
          <LicenseBlurb license={license} details={details} />
          <div className="x-small mt-3">{licenseDescription}</div>
        </div>
        {license === LicenseTypes.creativeCommons && (
          <Hyperlink
            className="text-primary-500 x-small"
            destination="https://creativecommons.org/about"
            target="_blank"
          >
            <FormattedMessage {...messages.viewLicenseDetailsLabel} />
          </Hyperlink>
        )}
      </Stack>
    );
  }
  return null;
};

LicenseDisplay.propTypes = {
  license: PropTypes.string.isRequired,
  details: PropTypes.shape({
    attribution: PropTypes.bool.isRequired,
    noncommercial: PropTypes.bool.isRequired,
    noDerivatives: PropTypes.bool.isRequired,
    shareAlike: PropTypes.bool.isRequired,
  }).isRequired,
  level: PropTypes.string.isRequired,
  licenseDescription: PropTypes.string.isRequired,
};

export default LicenseDisplay;
