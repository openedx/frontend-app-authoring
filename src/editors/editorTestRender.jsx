import React from 'react';
import { Provider } from 'react-redux';
import { render as baseRender } from '../testUtils';
import { EditorContextProvider } from './EditorContext';
import editorStore from './data/store';

/**
 * Custom render function for testing React components with the editor context and Redux store.
 *
 * Wraps the provided UI in both the EditorContextProvider and Redux Provider,
 * ensuring that components under test have access to the necessary context and store.
 *
 * @param {React.ReactElement} ui - The React element to render.
 * @param {string} [learningContextId='course-v1:Org+COURSE+RUN'] - Optional learning context ID
 * for the EditorContextProvider.
 * @returns {RenderResult} The result of the render, as returned by RTL render.
 */
const editorRender = (ui, learningContextId = 'course-v1:Org+COURSE+RUN') => baseRender(ui, {
  extraWrapper: ({ children }) => (
    <EditorContextProvider learningContextId={learningContextId}>
      <Provider store={editorStore}>
        {children}
      </Provider>
    </EditorContextProvider>
  ),
});

export default editorRender;
