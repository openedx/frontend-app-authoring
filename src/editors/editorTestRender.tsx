import React from 'react';
import { Provider } from 'react-redux';
import { type Store } from 'redux';
import { render as baseRender, type RouteOptions } from '../testUtils';
import { EditorContextProvider } from './EditorContext';
import { createStore } from './data/store';
import { EditorState } from './data/redux';

/**
 * Partial<EditorState> only allows top-level keys to be missing. This is an
 * even more partial state that allows sub-keys to be missing.
 */
export type PartialEditorState = {
  [P in keyof EditorState]?: Partial<EditorState[P]> | undefined;
};

interface Options {
  learningContextId?: string;
  initialState?: PartialEditorState;
}

let editorStore: Store<EditorState>;

/**
 * Custom render function for testing React components with the editor context and Redux store.
 *
 * Wraps the provided UI in both the EditorContextProvider and Redux Provider,
 * ensuring that components under test have access to the necessary context and store.
 *
 * @param {React.ReactElement} ui - The React element to render.
 * @param options - Options
 * @returns {RenderResult} The result of the render, as returned by RTL render.
 */
const editorRender = (ui, {
  learningContextId = 'course-v1:Org+COURSE+RUN',
  initialState = undefined,
  ...routerOptions
}: Options & RouteOptions = {}) => {
  editorStore = createStore(initialState);
  return baseRender(ui, {
    extraWrapper: ({ children }) => (
      <EditorContextProvider learningContextId={learningContextId}>
        <Provider store={editorStore}>
          {children}
        </Provider>
      </EditorContextProvider>
    ),
    ...routerOptions,
  });
};

export function getEditorStore() {
  return editorStore;
}

export default editorRender;
