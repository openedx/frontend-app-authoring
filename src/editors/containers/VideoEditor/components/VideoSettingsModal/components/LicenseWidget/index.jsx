import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  FormattedMessage,
  useIntl,
} from '@edx/frontend-platform/i18n';
import {
  Button,
  Stack,
} from '@openedx/paragon';
import { Add } from '@openedx/paragon/icons';

import { actions, selectors } from '../../../../../../data/redux';
import * as hooks from './hooks';
import messages from './messages';
import CollapsibleFormWidget from '../CollapsibleFormWidget';
import LicenseBlurb from './LicenseBlurb';
import LicenseSelector from './LicenseSelector';
import LicenseDetails from './LicenseDetails';
import LicenseDisplay from './LicenseDisplay';

/**
 * Collapsible Form widget controlling video license type and details
 */
const LicenseWidget = ({
  // redux
  isLibrary,
  licenseType,
  licenseDetails,
  courseLicenseType,
  courseLicenseDetails,
  updateField,
}) => {
  const intl = useIntl();
  const { license, details, level } = hooks.determineLicense({
    isLibrary,
    licenseType,
    licenseDetails,
    courseLicenseType,
    courseLicenseDetails,
  });
  const { licenseDescription, levelDescription } = hooks.determineText({ level });
  return (
    <CollapsibleFormWidget
      subtitle={(
        <div>
          <LicenseBlurb license={license} details={details} />
          <div className="x-small mt-2">{levelDescription}</div>
        </div>
      )}
      title={intl.formatMessage(messages.title)}
    >
      <Stack gap={4}>
        {license ? (
          <>
            <LicenseSelector license={license} level={level} />
            <LicenseDetails license={license} details={details} level={level} />
            <LicenseDisplay
              license={license}
              details={details}
              licenseDescription={licenseDescription}
            />
          </>
        ) : null }
        {!licenseType ? (
          <>
            <div className="border-primary-100 border-bottom my-2" />
            <Button
              className="text-primary-500 font-weight-bold justify-content-start pl-0"
              size="sm"
              iconBefore={Add}
              variant="link"
              onClick={() => updateField({ licenseType: 'select', licenseDetails: {} })}
            >
              <FormattedMessage {...messages.addLicenseButtonLabel} />
            </Button>
          </>
        ) : null }
      </Stack>
    </CollapsibleFormWidget>
  );
};

LicenseWidget.propTypes = {
  // redux
  isLibrary: PropTypes.bool.isRequired,
  licenseType: PropTypes.string.isRequired,
  licenseDetails: PropTypes.shape({}).isRequired,
  courseLicenseType: PropTypes.string.isRequired,
  courseLicenseDetails: PropTypes.shape({}).isRequired,
  updateField: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  isLibrary: selectors.app.isLibrary(state),
  licenseType: selectors.video.licenseType(state),
  licenseDetails: selectors.video.licenseDetails(state),
  courseLicenseType: selectors.video.courseLicenseType(state),
  courseLicenseDetails: selectors.video.courseLicenseDetails(state),
});

export const mapDispatchToProps = (dispatch) => ({
  updateField: (stateUpdate) => dispatch(actions.video.updateField(stateUpdate)),
});

export const LicenseWidgetInternal = LicenseWidget; // For testing only
export default connect(mapStateToProps, mapDispatchToProps)(LicenseWidget);
