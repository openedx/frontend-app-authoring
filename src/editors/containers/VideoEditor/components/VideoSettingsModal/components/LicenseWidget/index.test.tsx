import React from 'react';
import { screen } from '@testing-library/react';
import { initializeMocks } from '@src/testUtils';
import { editorRender, type PartialEditorState } from '@src/editors/editorTestRender';
import LicenseWidget from '.';

describe('LicenseWidget', () => {
  const baseState: PartialEditorState = {
    app: {
      learningContextId: 'random-id', // not a library ID, so isLibrary returns false
      blockId: null,
    },
    video: {
      licenseType: '',
      licenseDetails: {},
      courseLicenseType: 'all-rights-reserved',
      courseLicenseDetails: {},
    },
    requests: {},
  };
  beforeEach(() => {
    initializeMocks({});
  });

  test('renders as expected with default props (no license)', () => {
    editorRender(<LicenseWidget />, { initialState: baseState });

    expect(screen.getByText('License')).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: 'Add a license for this video' }),
    ).toBeInTheDocument();
  });

  test('renders without add button when isLibrary is true', () => {
    const isLibraryState = {
      ...baseState,
      app: {
        ...baseState.app,
        learningContextId: 'library-v1:abc123', // will trigger isLibrary selector
      },
      video: {
        ...baseState.video,
        licenseType: 'all-rights-reserved',
      },
    };

    editorRender(<LicenseWidget />, { initialState: isLibraryState });

    expect(screen.getByText('License')).toBeInTheDocument();
    expect(
      screen.queryByRole('button', { name: 'Add a license for this video' }),
    ).not.toBeInTheDocument();
  });

  test('renders as expected when licenseType is defined', () => {
    const licenseTypeState = {
      ...baseState,
      video: {
        ...baseState.video,
        licenseType: 'all-rights-reserved',
      },
    };

    editorRender(<LicenseWidget />, { initialState: licenseTypeState });

    expect(screen.getByText('License')).toBeInTheDocument();
    expect(
      screen.queryByRole('button', { name: 'Add a license for this video' }),
    ).not.toBeInTheDocument();
  });
});
