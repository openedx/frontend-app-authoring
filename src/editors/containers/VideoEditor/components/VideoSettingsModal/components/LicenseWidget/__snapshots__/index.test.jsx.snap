// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseWidget snapshots snapshots: renders as expected with default props 1`] = `
<injectIntl(ShimmedIntlComponent)
  subtitle={
    <div>
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        license="all-rights-reserved"
      />
      <div
        className="x-small mt-2"
      >
        <FormattedMessage
          defaultMessage="This license currently set at the course level"
          description="Helper text for license type when using course license"
          id="authoring.videoeditor.license.courseLevelDescription.helperText"
        />
      </div>
    </div>
  }
  title="License"
>
  <Stack
    gap={4}
  >
    <Fragment>
      <injectIntl(ShimmedIntlComponent)
        level="course"
        license="all-rights-reserved"
      />
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        level="course"
        license="all-rights-reserved"
      />
      <LicenseDisplay
        details={{}}
        license="all-rights-reserved"
        licenseDescription={
          <FormattedMessage
            defaultMessage="Licenses set at the course level appear at the bottom of courseware pages within your course."
            description="Message explaining where course level licenses are set"
            id="authoring.videoeditor.license.courseLicenseDescription.message"
          />
        }
      />
    </Fragment>
    <Fragment>
      <div
        className="border-primary-100 border-bottom my-2"
      />
      <Button
        className="text-primary-500 font-weight-bold justify-content-start pl-0"
        onClick={[Function]}
        size="sm"
        variant="link"
      >
        <FormattedMessage
          defaultMessage="Add a license for this video"
          description="Label for add license button"
          id="authoring.videoeditor.license.add.label"
        />
      </Button>
    </Fragment>
  </Stack>
</injectIntl(ShimmedIntlComponent)>
`;

exports[`LicenseWidget snapshots snapshots: renders as expected with isLibrary true 1`] = `
<injectIntl(ShimmedIntlComponent)
  subtitle={
    <div>
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        license="all-rights-reserved"
      />
      <div
        className="x-small mt-2"
      >
        <FormattedMessage
          defaultMessage="This license currently set at the library level"
          description="Helper text for license type when using library license"
          id="authoring.videoeditor.license.libraryLevelDescription.helperText"
        />
      </div>
    </div>
  }
  title="License"
>
  <Stack
    gap={4}
  >
    <Fragment>
      <injectIntl(ShimmedIntlComponent)
        level="library"
        license="all-rights-reserved"
      />
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        level="library"
        license="all-rights-reserved"
      />
      <LicenseDisplay
        details={{}}
        license="all-rights-reserved"
        licenseDescription={
          <FormattedMessage
            defaultMessage="Licenses set at the library level appear at the specific library video."
            description="Message explaining where library level licenses are set"
            id="authoring.videoeditor.license.libraryLicenseDescription.message"
          />
        }
      />
    </Fragment>
  </Stack>
</injectIntl(ShimmedIntlComponent)>
`;

exports[`LicenseWidget snapshots snapshots: renders as expected with licenseType defined 1`] = `
<injectIntl(ShimmedIntlComponent)
  subtitle={
    <div>
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        license="all-rights-reserved"
      />
      <div
        className="x-small mt-2"
      >
        <FormattedMessage
          defaultMessage="This license is set specifically for this video"
          description="Helper text for license type when choosing for a spcific video"
          id="authoring.videoeditor.license.defaultLevelDescription.helperText"
        />
      </div>
    </div>
  }
  title="License"
>
  <Stack
    gap={4}
  >
    <Fragment>
      <injectIntl(ShimmedIntlComponent)
        level="block"
        license="all-rights-reserved"
      />
      <injectIntl(ShimmedIntlComponent)
        details={{}}
        level="block"
        license="all-rights-reserved"
      />
      <LicenseDisplay
        details={{}}
        license="all-rights-reserved"
        licenseDescription={
          <FormattedMessage
            defaultMessage="When a video has a different license than the course as a whole, learners see the license at the bottom right of the video player."
            description="Message explaining where video specific licenses are seen by users"
            id="authoring.videoeditor.license.defaultLicenseDescription.message"
          />
        }
      />
    </Fragment>
  </Stack>
</injectIntl(ShimmedIntlComponent)>
`;
